src/core/tensor/tensor.cc

```diff
@@ -1050,6 +1135,42 @@ void Mult(const SType alpha, const Tensor &A, const Tensor &B, const SType beta,
   }
 }
 
+  // ************************
+  // Some traversal operations (works on const declarations without modifying tensor variables)
+  // ************************
+
+  //this function checks whether the next index falls on a special multiplier of the outer shape
+  //so the algorithm knows when to jump over/back to a starting element of the outer shape
+  //for e.g. in [[1,4,7], [2,5,8], [3,6,9]], elements 1,2,3 are the starting elements of their respective rows
+  //this additional check only has 1 loop for 2d matrix
+  //but runtime performance might degrade to O(nlog(n)) for higher dimensional tensors
+  int determine_order(int counter) const {
+      for (int n=0; n<(shape_multiplier_.size()-1); ++n) {
+          if((counter%shape_multiplier_[n])==0){
+              return ((shape_multiplier_.size()) - 1 - n);
+          }
+      }
+      return 0;
+  };
+
+  //this function updates the base indexes with the current index after every single traversal step, can be generalized beyond 2d cases
+  void update_base_index(std::vector<int>& traversal_info) const {
+      for (int n=0; n<(traversal_info[3]+1); ++n) {
+          traversal_info[n] = traversal_info[2];
+      }
+  };
+
+  //function to traverse a const strided tensor object
+  //it requires an additional vector, traversal_info {0,0,0,0}, comprising 4 elements
+  //index 0 and 1 store the base row and column index respectively (for 2d matrix)
```

Comment:
does it mean that the traverse_next function only work for 2d matrix?

Reply:
this function can work for n-dimensional cases if there are more elements in the traversal_info vector, but in this example, it only works for 2d matrix